vectordb_orm/query.py

```diff
@@ -0,0 +1,142 @@
+from vectordb_orm.attributes import AttributeCompare
+from vectordb_orm.results import QueryResult
+from vectordb_orm.fields import EmbeddingField
+from vectordb_orm.base import MilvusBase
+from typing import Any
+
+# https://milvus.io/docs/search.md
+MAX_MILVUS_INT = 16384
+
+class MilvusQueryBuilder:
+    """
+    Recursive query builder to allow for chaining of queries.
+
+    Example:
+        >>> query(MyObject).filter(MyObject.text == "hello").limit(10).all()
+        >>> query(MyObject.text).filter(MyObject.text == "hello").limit(10).all()
+        >>> query(MyObject.text, MyObject.embedding).filter(MyObject.text == "hello").limit(10).all()
+
+    """
+    def __init__(
+        self,
+        milvus_client: Milvus,
+        *query_objects: MilvusBase | AttributeCompare
+    ):
+        if len(query_objects) == 0:
+            raise ValueError("Must specify at least one query object.")
+
+        self.milvus_client = milvus_client
+        if isinstance(query_objects[0], AttributeCompare):
+            self.cls = query_objects[0].base_cls
+        elif issubclass(query_objects[0], MilvusBase):
+            self.cls = query_objects[0]
+        else:
+            raise ValueError(f"Query object must be of type `MilvusBase` or `AttributeCompare`, got {type(query_objects[0])}")
+
+        # Queries created over time
+        self._query_objects = list(query_objects)
+        self._filters: list[str] = []
+        self._limit : int | None = None
+        self._offset : int | None = None
+
+        # Set if similarity ranking is used
+        self._similarity_attribute : AttributeCompare | None = None
+        self._similarity_value = None
+
+    def filter(self, *filters: AttributeCompare):
+        for f in filters:
+            self._filters.append(f.to_expression())
+        return self
+
+    def offset(self, offset: int):
+        self._offset = offset
+        return self
+
+    def limit(self, limit: int | None):
+        self._limit = limit
+        return self
+
+    def order_by_similarity(self, embedding_attr: AttributeCompare, query_embedding):
+        if self._similarity_attribute is not None:
+            raise ValueError("Only one similarity ordering can be used per query.")
+        self._similarity_attribute = embedding_attr
+        self._similarity_value = query_embedding
+        return self
+
+    def all(self):
+        filters = " and ".join(self._filters) if self._filters else None
+        query_records = [self._similarity_value] if self._similarity_value is not None else None
+        embedding_field_name = self._similarity_attribute.attr if self._similarity_attribute else None
+        output_fields = self._get_output_fields()
+
+        offset = self._offset if self._offset is not None else 0
+        # Sum of limit and offset should be less than MAX_MILVUS_INT
+        limit = self._limit if self._limit is not None else (MAX_MILVUS_INT - offset)
+
+        params = {"nprobe": 16}
+
+        if query_records:
```

Comment:
for the uninitiated, may be good to semantically distinguish between a Milvus query vs search

from the Github [repo](https://github.com/milvus-io/pymilvus/blob/a1e3734dada1644e4ed0774e1408b8bf74836732/pymilvus/client/stub.py#L901) I see this distinction:

Query: `Query with a set of criteria, and results in a list of records that match the query exactly.`
Search:  `Searches a collection based on the given expression and returns query results.`

They both seem to just call `handler.query` with different arguments similar to how you have it so it may not be worth a refactor, but from a code perspective this section looks a bit all-powerful. adding comments on this is probably sufficient